### DISCUSSION TITLE
B2B with multiple instances share opaque token validation across all instances

### DISCUSSION THREAD

**Discussion by stiwari99** [zitadel/zitadel#7140] [General]
In the SaaS model, when dealing with B2B customers who might have their own set of B2B clients or individual users, it's understood that separate instances must be set up for each B2B customer as per the documentation.

Given that there's a singular SaaS service, a project needs to be established within each instance to facilitate logins. We utilize an opaque token for the login process, which subsequently needs validation within the SaaS through an introspection endpoint.

For every project instance, a unique JWT private key must be generated and stored within the SaaS platform. This ensures the ability to validate the opaque token specific to each instance.

My query is: Is there an alternative method where a single JWT private key could be stored on the backend, enabling the validation of opaque tokens across all instances?

Also, is it possible to share saas project across all instances so that we don't have to create it with every b2b customer

**Comment by hifabienne** [comment]
hei @livio-a @muhlemmer can you help here?

**Reply by stiwari99** [reply]
Any updates on this?

**Reply by muhlemmer** [reply]
Instances are isolated objects and you cannot share tokens or projects among them. Use organizations instead, those are also suitable for B2B use cases and allows applications and users to receive grants from other organizations.

If you insist using separate instances, it might fit your usecase to use  one instance as Generic OIDC provider for the other instance, but that will be much more complex than just using orgs.

**Reply by stiwari99** [reply]
Hello @muhlemmer,

Thank you for your response. I'm interested in understanding how I can implement the following scenario (See attached diagram):

Within our SaaS platform, let's consider two organizations: Pepsi and Coke. Each of these organizations has its own roster of employees with distinct email addresses, such as shane@pepsi.com and rob@coke.com.

Both Pepsi and Coke serve individual end-users, represented by email addresses like alice@gmail.com. The objective is for alice@gmail.com to access both the Pepsi and Coke platforms using her single email and view the respective resources on each.

Moreover, Pepsi and Coke could potentially have another organization, such as Water LLC, as their client. In this case, users from Water LLC should be able to log into both the Pepsi and Coke platforms to access their resources.

Currently, we have a custom solution to facilitate this functionality. However, we are considering transitioning to Zitadel. I'm uncertain about how to replicate this specific use-case with Zitadel.

Could you please advise on how to enable users like Alice to log in to both Pepsi and Coke using a shared email address? I'm particularly interested in understanding if this can be accomplished through either of the following methods:

1. Having a unique email address across all instances.
2. Assigning a unique email address within each organization.
![Screen Shot 2024-01-08 at 11 51 20 AM](https://github.com/zitadel/zitadel/assets/585384/86b7abba-e60b-4e87-a74e-851e677b676d)

**Question**

1. is it possible to login via email or phone number in an organization where email and phone numbers are unique in an org and not across instances?
2. Do we have any documentation or how this can be done `use one instance as Generic OIDC provider for the other instance`

**Reply by hifabienne** [reply]
hei @bigbang4u2 
One important thing to understand is the difference between authentication and authorization.
So a user within ZITADEL does always exist in an Organization. Think of it as the owner of the user. The one who manages the user.
But the user is allowed to have permissions for multiple organizations.
So you can have the following organizations:
Pepsi, Coke, End Users.

The endusers live in the end users organization, but can have permissions to other orgnizations.

I think the following guides can help you:
B2B Scenario: https://zitadel.com/docs/guides/solution-scenarios/b2b
Use email to login: https://zitadel.com/docs/guides/solution-scenarios/configurations#use-email-to-login

**Reply by stiwari99** [reply]
@hifabienne thanks for response, but I see one issue with creating end user organization.

End User will go to `pepsi.com` and will perform registration say with email alice@gmail.com
End user should also be able to register on `coke.com` with email alice@gmail.com

As an end user they are not aware of saas service provider they only know about pepsi.com and coke.com as two different companies where they can register themselves

**Reply by muhlemmer** [reply]
@bigbang4u2 that is completely fine. An email can be used for multiple user accounts and does [not need to be unique](https://zitadel.com/docs/concepts/structure/users#constraints) within one instance:

> User accounts are uniquely identified by their id or loginname in combination of the organization domain (eg, road.runner@acme.zitadel.local). You can use the same email address for different user accounts.

Not meaning to be impolite but might I suggest reading the documentation first? We have put extensive work in documenting most use-cases and concepts already. You can start with the solution scenarios Fabienne posted and be sure to read https://zitadel.com/docs/concepts to understand how we model instances, organizations, users, projects, domains and applications.

**Reply by stiwari99** [reply]
Hi muhlemmer, 
Thanks for prompt response, I did over documentation and did some POC that is where I encountered this problem. 

I agree that user can register in both pepsi.com and coke.com with email alice@gmail.com (only if `Add organization domain as suffix to loginnames ` settings is enabled for an org). In order to use alice@gmail.com as login we have to pass org id in scope


Key piece was missing from login with gmail email was to add  scope: `'openid profile email urn:zitadel:iam:org:id:64426769854333334' so that login will automatically be directed to organization login rather than generic login page where we have to provide `username@subdomain` to login into particular org

```
const authConfig = {
    authority: 'https://myinstance-tdderdd.zitadel.cloud', //Replace with your issuer URL
    client_id: '434244344988359@portal', //Replace with your client id
    redirect_uri: 'http://localhost:3000/callback',
    response_type: 'code',
    scope: 'openid profile email urn:zitadel:iam:org:id:64426769854333334',
    post_logout_redirect_uri: 'http://localhost:3000',
    userinfo_endpoint: 'https://myinstance-tdderdd.zitadel.cloud/oidc/v1/userinfo', //Replace with your user-info endpoint
    response_mode: 'query',
    code_challenge_method: 'S256',
  };

```
So for pepsi we have to pass pepsi orgId in scope. Similarly we can do for coke
This way login will be automatically redirected to respective organization

**Is there any recommended way how to dynamically add org id to scope based on if login.pepsi.com or login.coke.com are hit, so that it allows login into specific organization  
One way I can think of creating mapping of domain and orgId mapping behind public endpoint and hit that endpoint to return orgId to update config**

I tried using primary domain (`urn:zitadel:iam:org:domain:primary:demo-vendor.myinstance-tdderdd.zitadel.cloud`) as scope but then it does not allow us to enter email instead asking to enter `username@demo-vendor.myinstance-tdderdd.zitadel.cloud`

Seems like its an open issue to link custom domain with corresponding org https://github.com/zitadel/zitadel/issues/5094

**Reply by stiwari99** [reply]
Any suggestions or workaround on how can we provide organization its own login url e.g `login-org1.example.com`, `login-org2.example.com` etc where they can login directly and it has organization scope e.g `urn:zitadel:iam:org:id:64426769854333334`  

Or 
Alternatively, if we can somehow map orgId with domain, that way we can pass it in scope during authorization
Do we need to keep orgId to subdomain mapping in frontend? If so then this means all of our orgs are publicly exposed

**Comment by stiwari99** [comment]
Something like this would be nice to have where we take organization name instead of id
https://auth0.com/docs/manage-users/organizations/configure-organizations/use-org-name-authentication-api

This way we don't have to keep mapping of org Id to org names or domain to pass in scopes
